flutter_vma/flutter_skia_vma.cc

```diff
@@ -0,0 +1,274 @@
+  if (kProtected_AllocationPropertyFlag & allocationPropertyFlags) {
+    info.requiredFlags |= VK_MEMORY_PROPERTY_PROTECTED_BIT;
+  }
+
+  VmaAllocation allocation;
+  VkResult result =
+      vmaAllocateMemoryForImage(allocator_, image, &info, &allocation, nullptr);
+  if (VK_SUCCESS == result) {
+    *backendMemory = reinterpret_cast<skgpu::VulkanBackendMemory>(allocation);
+  }
+  return result;
+}
+
+VkResult FlutterSkiaVulkanMemoryAllocator::allocateBufferMemory(
+    VkBuffer buffer,
+    BufferUsage usage,
+    uint32_t allocationPropertyFlags,
+    skgpu::VulkanBackendMemory* backendMemory) {
+  VmaAllocationCreateInfo info;
+  info.flags = 0;
+  info.usage = VMA_MEMORY_USAGE_UNKNOWN;
+  info.memoryTypeBits = 0;
+  info.pool = VK_NULL_HANDLE;
+  info.pUserData = nullptr;
+
+  switch (usage) {
+    case BufferUsage::kGpuOnly:
+      info.requiredFlags = VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT;
+      info.preferredFlags = 0;
+      break;
+    case BufferUsage::kCpuWritesGpuReads:
+      // When doing cpu writes and gpu reads the general rule of thumb is to use
+      // coherent memory. Though this depends on the fact that we are not doing
+      // any cpu reads and the cpu writes are sequential. For sparse writes we'd
+      // want cpu cached memory, however we don't do these types of writes in
+      // Skia.
+      //
+      // TODO (kaushikiska): In the future there may be times where specific
+      // types of memory could benefit from a coherent and cached memory.
+      // Typically these allow for the gpu to read cpu writes from the cache
+      // without needing to flush the writes throughout the cache. The reverse
+      // is not true and GPU writes tend to invalidate the cache regardless.
+      // Also these gpu cache read access are typically lower bandwidth than
+      // non-cached memory. For now Skia doesn't really have a need or want of
+      // this type of memory. But if we ever do we could pass in an
+      // AllocationPropertyFlag that requests the cached property.
+      info.requiredFlags = VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT |
+                           VK_MEMORY_PROPERTY_HOST_COHERENT_BIT;
+      info.preferredFlags = VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT;
+      break;
+    case BufferUsage::kTransfersFromCpuToGpu:
+      info.requiredFlags = VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT |
+                           VK_MEMORY_PROPERTY_HOST_COHERENT_BIT;
+      info.preferredFlags = 0;
+      break;
+    case BufferUsage::kTransfersFromGpuToCpu:
+      info.requiredFlags = VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT;
+      info.preferredFlags = VK_MEMORY_PROPERTY_HOST_CACHED_BIT;
+      break;
+  }
+
+  if (must_use_coherent_host_visible_memory_ &&
+      (info.requiredFlags & VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT)) {
+    info.requiredFlags |= VK_MEMORY_PROPERTY_HOST_COHERENT_BIT;
+  }
+  if (kDedicatedAllocation_AllocationPropertyFlag & allocationPropertyFlags) {
+    info.flags |= VMA_ALLOCATION_CREATE_DEDICATED_MEMORY_BIT;
```

Comment:
This reminds me, perhaps we should do this for all images from the image decoder (the ui.Images)?

Reply:
Noted, will fix this in `allocator_vk.cc`